File: go-leet-code/84. Largest Rectangle in Histogram.go
package main

import (
	"fmt"
)

func top(stack []int) int {
	return stack[len(stack)-1]
}
func pop(st []int) []int {
	st = st[:len(st)-1]
	return st

}
func largestRectangleArea(heights []int) int {
	mx, ret := 0, 0

	st := make([]int, 0)
	n := len(heights)
	for i := 0; i <= n; i++ {

		now := 0
		if i < n {
			now = heights[i]
		}

		for len(st) > 0 && heights[top(st)] > now {
			j := top(st)
			st = pop(st)
			k := -1
			if len(st) > 0 {
				k = top(st)
			}
			ret = (i - k - 1) * heights[j]
			if ret > mx {
				mx = ret
			}
			//fmt.Printf("i:%d j:%d k:%d ret:%d\n",i,j,k,ret)
		}
		st = append(st, i)
	}
	return mx
}

func main() {
	var arr = []int{2, 2, 2}
	fmt.Println(largestRectangleArea(arr))
}

File: Leetcode practice/84. Largest Rectangle in Histogram.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
using namespace std;


class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        // printf("height: { ");
        // for(int i = 0;i<heights.size();i++) {
        //     printf(" %d",heights[i]);
        // }
        // printf(" }\n");
        stack<int> st;
        int ret = 0, mx = 0, n = heights.size();
        int h;
        for(int i = 0;i<=n;i++){
            if (i == n) h = 0;
            else h = heights[i];
            while(st.size() > 0 && heights[st.top()] > h) {
                int t = st.top();
                st.pop();
                ret = heights[t] * (st.size() ? i - st.top() - 1 : i);
                mx = max(mx,ret);
            }
            st.push(i);
        }
        return mx;        
    }
};


int main() {
    vector< vector<int> > tests = {
        {1,2,3,4,3,2,1},
        {1,2,3,4},
        {1,1,2,3,1,2,2,2},
        {1,0,3}
    };

    Solution sol;
    for(auto vec : tests) {
        printf("%d\n", sol.largestRectangleArea(vec));
    }
    // vector<int> v1 ={1,0,3};
    // printf("%d\n",sol.largestRectangleArea(v1));

}